lib/components/form/connected-links.js

```diff
@@ -0,0 +1,39 @@
+import React from 'react'
+import { connect } from 'react-redux'
+import { LinkContainer } from 'react-router-bootstrap'
+import { Redirect } from 'react-router'
+import { Link } from 'react-router-dom'
+
+/**
+ * This function enhances the routing components imported above
+ * by preserving the itinerary search query from the redux state
+ * when redirecting the user between the main map and account-related pages,
+ * so that when the user returns to the map, the itinerary that was previously
+ * displayed is shown again.
+ * Implementers only need to specify the 'to' route and
+ * do not need to hook to redux store to retrieve the itinerary search query.
+ * @param RoutingComponent The routing component to enhance.
+ * @returns A new component that passes the redux search params to
+ *          the RoutingComponent's 'to' prop.
+ */
+const connectLink = RoutingComponent =>
+  ({ children, queryParams, to, ...props }) => (
+    <RoutingComponent {...props} to={{ pathname: to, search: queryParams }}>
+      {children}
+    </RoutingComponent>
+  )
+
+// For connecting to the redux store
+const mapStateToProps = (state, ownProps) => {
+  return {
+    queryParams: state.router.location.search
+  }
+}
+
+// Enhance routing components, connect the result to redux,
+// and export.
+export default {
+  Link: connect(mapStateToProps)(connectLink(Link)),
+  LinkContainer: connect(mapStateToProps)(connectLink(LinkContainer)),
+  Redirect: connect(mapStateToProps)(connectLink(Redirect))
```

Comment:
I think these component names would benefit from having some mention of their query param-preserving behavior. E.g.:
```
LinkWithQuery
LinkContainerWithQuery
RedirectWithQuery
```
I'm open to other suggestions for names, but I think it would be good to somehow distinguish them from their vanilla counterparts in usage.

Reply:
Updated in https://github.com/opentripplanner/otp-react-redux/pull/280/commits/35eb848eeb28b878ba59d7fd07e01b18f92fa260.